component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/OneTimePassword.java

```diff
@@ -148,23 +148,82 @@ public static String generateOTP(byte[] secret, long movingFactor, int codeDigit
+     * @throws NoSuchAlgorithmException if no provider makes either HmacSHA1 or HMAC-SHA-1 digest
+     *                                  algorithms available.
+     * @throws InvalidKeyException      The secret provided was not a valid HMAC-SHA-1 key.
+     */
+    public static String generateAlphaNumericOTP(byte[] secret, long movingFactor, int codeDigits, boolean addChecksum,
+                                                 int truncationOffset) throws NoSuchAlgorithmException, InvalidKeyException {
+        // put movingFactor value into text byte array
+        String result = null;
+        int digits = addChecksum ? (codeDigits + 1) : codeDigits;
+        byte[] text = new byte[8];
+        for (int i = text.length - 1; i >= 0; i--) {
+            text[i] = (byte) (movingFactor & 0xff);
+            movingFactor >>= 8;
+        }
+        // compute hmac hash
+        byte[] hash = hmacShaGenerate(secret, text);
+        // put selected bytes into result int
+        int offset = hash[hash.length - 1] & 0xf;
+        if ((0 <= truncationOffset) && (truncationOffset < (hash.length - 8))) {
+            offset = truncationOffset;
+        }
+        int firstBinary = ((hash[offset] & 0x7f) << 24) | ((hash[offset + 1] & 0xff) << 16)
+                | ((hash[offset + 2] & 0xff) << 8) | ((hash[offset + 3] & 0xff));
+        int secondBinary = ((hash[offset + 4] & 0x7f) << 24) | ((hash[offset + 5] & 0xff) << 16)
+                | ((hash[offset + 6] & 0xff) << 8) | ((hash[offset + 7] & 0xff));
+        result = Integer.toString(firstBinary, 36).concat(Integer.toString(secondBinary, 36)).toUpperCase();
+        while (result.length() < digits) {
+            result = "A" + result;
+        }
+        result = result.substring(result.length() - digits, result.length());
+        return result;
+    }
+
     /**
      * Generate the token.
      *
      * @param key    the key
      * @param base   the base
      * @param digits the number of digits
+     * @param isEnableAlphanumericToken a flag that indicates the token is alphanumeric or not
      * @return the generated token
      */
-    public String generateToken(String key, String base, int digits) {
+    public String generateToken(String key, String base, int digits, boolean isEnableAlphanumericToken) {
         int truncOffset = 0;
-        truncOffset = 0;
-        try {
-            return generateOTP(key.getBytes(), Long.parseLong(base), digits, false, truncOffset);
-        } catch (NoSuchAlgorithmException e) {
-            log.error("Unable to find the Algorithm", e);
-        } catch (InvalidKeyException e) {
-            log.error("Unable to find the secret key", e);
+        if (isEnableAlphanumericToken){
+            try {
+                return generateAlphaNumericOTP(key.getBytes(), Long.parseLong(base), digits, false, truncOffset);
+            } catch (NoSuchAlgorithmException e) {
+                log.error("Unable to find the Algorithm", e);
```

Comment:
Why are we not throwing the exception? Shall we add some error specific properties May algorithm name